api/src/main/java/org/motechproject/nms/api/web/UserController.java

```diff
@@ -31,47 +39,123 @@
 
         ResponseUser user = null;
 
-        if (MOBILE_ACADEMY.equals(serviceName)) {
-            user = new MobileAcademyUser();
-        } else if (MOBILE_KUNJI.equals(serviceName)) {
-            user = new MobileKunjiUser();
-        } else if (KILKARI.equals(serviceName)) {
-            user = new KilkariResponseUser();
-            Subscriber subscriber = kilkariService.getSubscriber(callingNumber);
-            if (subscriber == null) {
-                throw new NotFoundException(String.format(NOT_FOUND, "callingNumber"));
-            }
-            Set<Subscription> subscriptions = subscriber.getSubscriptions();
-            Set<String> packs = new HashSet<>();
-            for (Subscription subscription : subscriptions) {
-                packs.add(subscription.getSubscriptionPack().getName());
-            }
-            ((KilkariResponseUser) user).setSubscriptionPackList(packs);
-        } else {
+        /*
+        Make sure the url the user hit corresponds to a service we are expecting
+         */
+        if (!(MOBILE_ACADEMY.equals(serviceName) || MOBILE_KUNJI.equals(serviceName)) && !(KILKARI.equals(serviceName))) {
             failureReasons.append(String.format(INVALID, "serviceName"));
         }
 
+        /*
+        Handle the FLW services
+         */
+        if (MOBILE_ACADEMY.equals(serviceName) || MOBILE_KUNJI.equals(serviceName)) {
+            user = getFrontLineWorkerResponseUser(serviceName, callingNumber);
+        }
+
+        /*
+        Kilkari in the house!
+         */
+        if (KILKARI.equals(serviceName)) {
+            user = getKilkariResponseUser(callingNumber);
+        }
+
         if (failureReasons.length() > 0) {
             throw new IllegalArgumentException(failureReasons.toString());
         }
 
-        if (languages.size() > 0) {
-            user.setLanguageLocationCode(languages.get(0).getCode());
+        Language language = languageService.getDefaultCircleLanguage(circle);
+        if (null != language) {
+            user.setDefaultLanguageLocationCode(language.getCode());
         }
 
         return user;
     }
+
+    private ResponseUser getKilkariResponseUser(String callingNumber) throws NotFoundException {
+        ResponseUser user;
+        user = new KilkariResponseUser();
+        Subscriber subscriber = kilkariService.getSubscriber(callingNumber);
+        if (subscriber == null) {
+            throw new NotFoundException(String.format(NOT_FOUND, "callingNumber"));
+        }
+        Set<Subscription> subscriptions = subscriber.getSubscriptions();
+        Set<String> packs = new HashSet<>();
+        for (Subscription subscription : subscriptions) {
+            packs.add(subscription.getSubscriptionPack().getName());
+        }
+        ((KilkariResponseUser) user).setSubscriptionPackList(packs);
+        return user;
+    }
+
+    private ResponseUser getFrontLineWorkerResponseUser(String serviceName, String callingNumber) {
+        ResponseUser user;
```

Comment:
Changing user to FrontLineWorkerUser will eliminate all the casting you're doing below.